File: ForumAPI/ForumAPI/Interfaces/GreenitorDAO.cs
using ForumAPI.DTOs;
using ForumAPI.DTOs.GreenitorDTOs;

namespace ForumAPI.Interfaces
{
    /// <summary>
    /// Interface defining data access operations related to Greenitors (users).
    /// </summary>
    public interface GreenitorDAO
    {
        /// <summary>
        /// Authenticates a user and generates a token for login.
        /// </summary>
        /// <param name="loginDTO">The login data transfer object containing the user's credentials.</param>
        /// <returns>A task representing the asynchronous operation, with a <see cref="TokenDTO"/> containing the login token.</returns>
        public Task<TokenDTO> Login(LoginDTO loginDTO);

        /// <summary>
        /// Registers a new user (Greenitor) in the system.
        /// </summary>
        /// <param name="registerUserDTO">The data transfer object containing the user's registration details, including their image.</param>
        /// <returns>A task representing the asynchronous operation, with a <see cref="ResponseMessage"/> indicating success or failure.</returns>
        public Task<ResponseMessage> RegisterUser(RegisterUserWithImageDTO registerUserDTO);

        /// <summary>
        /// Retrieves a user by their username.
        /// </summary>
        /// <param name="username">The username of the user to retrieve.</param>
        /// <returns>A task representing the asynchronous operation, with a <see cref="GreenitorDTO"/> containing the user's details.</returns>
        public Task<GreenitorDTO> GetUserByUsername(string username);

        /// <summary>
        /// Increments the interaction count for a specific user.
        /// </summary>
        /// <param name="username">The username of the user whose interactions count is to be incremented.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        Task IncrementUserInteractions(string username);

        /// <summary>
        /// Decrements the interaction count for a specific user.
        /// </summary>
        /// <param name="username">The username of the user whose interactions count is to be decremented.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        Task DecrementUserInteractions(string username);

        /// <summary>
        /// Retrieves a list of all Greenitors (users) without their roles.
        /// </summary>
        /// <returns>A task representing the asynchronous operation, with a list of <see cref="GreenitorWithoutRoleDTO"/> representing all users.</returns>
        Task<List<GreenitorWithoutRoleDTO>> GetAllGreenitors();
    }
}

File: ClientAPI/src/main/java/pt/upskill/clientapi/Interfaces/GreenitorDAO.java
package pt.upskill.clientapi.Interfaces;

import pt.upskill.clientapi.DTOs.LoginDTO;
import pt.upskill.clientapi.DTOs.ResponseMessage;
import pt.upskill.clientapi.DTOs.UpdateGreenitorDTO;
import pt.upskill.clientapi.Models.Greenitor;
import pt.upskill.clientapi.Models.Token;

import javax.security.auth.login.AccountNotFoundException;
import java.util.List;

/**
 * Interface for Data Access Object (DAO) operations related to the Greenitor entity.
 * Provides methods for managing Greenitor entities and their interactions.
 */
public interface GreenitorDAO {

    /**
     * Registers a new Greenitor (user).
     *
     * @param greenitor The Greenitor object containing the user's data.
     * @return The created Greenitor object.
     */
    Greenitor registerUser(Greenitor greenitor);

    /**
     * Logs in a Greenitor based on the provided login credentials.
     *
     * @param loginDTO The login data transfer object containing the user's login credentials.
     * @return A Token object containing the authentication token for the Greenitor.
     * @throws AccountNotFoundException if the user is not found.
     */
    Token loginUser(LoginDTO loginDTO);

    /**
     * Fetches a Greenitor by their username.
     *
     * @param username The username of the Greenitor.
     * @return The Greenitor object corresponding to the given username.
     */
    Greenitor getGreenitorByUsername(String username);

    /**
     * Increments the number of interactions for a Greenitor identified by their username.
     *
     * @param username The username of the Greenitor whose interactions need to be incremented.
     * @return A ResponseMessage indicating the result of the operation.
     */
    ResponseMessage incrementInteractions(String username);

    /**
     * Decrements the number of interactions for a Greenitor identified by their username.
     *
     * @param username The username of the Greenitor whose interactions need to be decremented.
     * @return A ResponseMessage indicating the result of the operation.
     */
    ResponseMessage decrementInteractions(String username);

    /**
     * Retrieves a list of all Greenitors.
     *
     * @return A list of Greenitor objects.
     */
    List<Greenitor> getAllGreenitors();

    /**
     * Updates the details of an existing Greenitor (user).
     *
     * @param username the username of the Greenitor to update
     * @param greenitor the updated user information
     * @return a ResponseMessage indicating the success of the update
     */
    ResponseMessage updateGreenitor(String username, UpdateGreenitorDTO greenitor);
}